### PR TITLE
Refactor the interface for actions updating states

### PR DESCRIPTION
**Pull request by dahlia**
This patch retouched `IAction` and `Transaction<T>` APIs to make library users easier to implement their own `IAction` classes.  Two key changes are:

 -  `Transaction<T>`'s `Sender`–`Recipient` model was replaced by `Signer`–`UpdatedAddresses` model.   Unlike cryptocurrencies, transactions in games are not necessarily a transfer of assets, so it is difficult to determine what type of assert is transferred or who will receives the asset.  A more useful perspective is, like what kind of transformation is performed, or what states are changed.

    To be close to this perspective, we decided to get rid of `Transaction<T>.Recipient` (which is singular) and have `Transaction<T>.UpdatedAddresses` (which is plural) instead. As there is no more asset to transfer, the term `Sender` was also renamed to `Signer`, which fits more to the new perspective.

 -  The `Address`es `IAction` tries to update no more need to be manually coded using `IAction.RequestStates()` method.  That method was removed at all, and updated `Address`es became automatically scanned using dirty-tracking for the most cases.

More detailed changes are written in *CHANGES.md* file.

----

### Tasks

- [x] `BlockChain` needs to validate if all `Transaction<T>.UpdatedAddresses` match to the actual addresses affected by `Transaction<T>.Actions`.   Note that as it depends on precedent states, cannot be done by `Transaction<T>` in itself.
- [x] `Block<T>.Validate()` had been not validating its `Transactions`.  `Transaction<T>.Validate()` also should be called inside that method.
- [x] Currently `UpdatedAddresses` must be coded by human when a `Transaction<T>` is instantiated.  There should be a façade factory to play the given actions and automatically determines the updated addresses.
- [ ] Get rid of the whole `AddressStateMap` thing?  It no differs from `IImmutableDictionary<Address, object>`.  See also #98.
- [x] Changelogs should be more organized and the whole change need to be explained at a higher level.

### CODE REVIEW

**Comment by dahlia** [comment]
@longfin Could you review this?

File: Libplanet.Tests/Blockchain/BlockChainTest.cs [hunk 1]
```diff
@@ -3,6 +3,7 @@
 using System.Collections.Immutable;
 using System.Drawing.Design;
 using System.Linq;
+using System.Runtime.CompilerServices;
```
Comment (longfin) [inline Libplanet.Tests/Blockchain/BlockChainTest.cs]: Is this using statement necessary?

File: Libplanet/Tx/Transaction.cs [hunk 2]
```diff
@@ -225,62 +236,147 @@ public static Transaction<T> FromBencodex(byte[] bytes)
...
+
+            if (!actionsArray.IsEmpty)
+            {
+                IAccountStateDelta delta = new Transaction<T>(
+                    signer,
+                    publicKey,
+                    updatedAddresses,
+                    ts,
+                    actionsArray
+                ).EvaluateActions(
```
Comment (longfin) [inline Libplanet/Tx/Transaction.cs]: It's important to point out that `IAction.Execute()` is being executed in `Transaction<T>.Create()`.

File: Libplanet.Tests/Blockchain/BlockChainTest.cs (hunk 1 above)
Comment (dahlia) [reply Libplanet.Tests/Blockchain/BlockChainTest.cs]: No, it's mistaken.  I will remove them!

File: Libplanet/Tx/Transaction.cs (hunk 2 above)
Comment (dahlia) [reply Libplanet/Tx/Transaction.cs]: It's for filling out `UpdatedAddresses` automatically, and documented in `<remarks>` section.

**Comment by dahlia** [comment]
@longfin I amended a commit to fix the build failure on Windows. Can you approve this again?

**Pull request merged**